tests/test_forward_ref.py

```diff
@@ -840,3 +859,99 @@ class Foo(BaseModel):
+    )
+
+
+#   NOTE: the `undefined_types_warning` tests below are "statically parameterized" (i.e. have Duplicate Code).
+#   The initial attempt to refactor them into a single parameterized test was not strateforward, due to the use of the
+#   `create_module` fixture and the need for `from __future__ import annotations` be the first line in a module.
+#
+#   Test Parameters:
+#     1. config setting: (1a) default behavior vs (1b) overriding with Config setting:
+#     2. type checking approach: (2a) `from __future__ import annotations` vs (2b) `ForwardRef`
+#
+#   The parameter tags "1a", "1b", "2a", and "2b" are used in the test names below, to indicate which combination
+#   of conditions the test is covering.
+
+
+def test_undefined_types_warning_1a_raised_by_default_2a_future_annotations(create_module):
+    with pytest_raises_undefined_types_warning(defining_class_name='Foobar', missing_type_name='UndefinedType'):
+        create_module(
+            # language=Python
+            """
+from __future__ import annotations
+from pydantic import BaseModel
+
+class Foobar(BaseModel):
+    a: UndefinedType
+"""
+        )
+
+
+def test_undefined_types_warning_1a_raised_by_default_2b_forward_ref(create_module):
+    with pytest_raises_undefined_types_warning(defining_class_name='Foobar', missing_type_name='UndefinedType'):
+
+        @create_module
+        def module():
+            from typing import ForwardRef
+
+            from pydantic import BaseModel
+
+            UndefinedType = ForwardRef('UndefinedType')
+
+            class Foobar(BaseModel):
+                a: UndefinedType
+
+
+def test_undefined_types_warning_1b_suppressed_via_config_2a_future_annotations(create_module):
+    module = create_module(
+        # language=Python
+        """
+from __future__ import annotations
+from pydantic import BaseModel
+
+class Foobar(BaseModel):
+    a: UndefinedType
+    # Suppress the undefined_types_warning
+    class Config:
+        undefined_types_warning = False
+"""
+    )
+    # Since we're testing the absence of a warning, it's important to confirm pydantic was actually run.
+    # The presence of the `__pydantic_model_complete__` is a good indicator of this.
+    assert module.Foobar.__pydantic_model_complete__ is False
+
+
+def test_undefined_types_warning_1b_suppressed_via_config_2b_forward_ref(create_module):
+    @create_module
+    def module():
+        from typing import ForwardRef
+
+        from pydantic import BaseModel
+
+        UndefinedType = ForwardRef('UndefinedType')
+
+        class Foobar(BaseModel):
+            a: UndefinedType
+
+            # Suppress the undefined_types_warning
+            class Config:
+                undefined_types_warning = False
+
+    # Since we're testing the absence of a warning, it's important to confirm pydantic was actually run.
+    # The presence of the `__pydantic_model_complete__` is a good indicator of this.
+    assert module.Foobar.__pydantic_model_complete__ is False
```

Comment:
@ybressler I've added a block of comments/docs at the top of these 4 tests to describe:

- why the tests have duplicated code (and some context to help someone untangle it the future)
- what the tests are doing (with respect to the parameters/conditions we're trying to cover).
- some simple tags for each parameter state that can be used in the test name (and provide a reference to the comment block in order to to avoid duplication of the notes/comments within the tests themselves).

Reply:
Excellent!